src/vcpkg/base/system.cpp

```diff
@@ -227,30 +228,32 @@ namespace vcpkg
 #endif
     }
 
-    void print_environment_variables()
+    std::string get_environment_variables()
     {
-        msg::write_unlocalized_text_to_stdout(Color::none, "The following environment variables are currently set:\n");
+        std::string result;
 #if defined(_WIN32)
-        auto free = [](LPWCH p) { FreeEnvironmentStringsW(p); };
-        std::unique_ptr<WCHAR, decltype(free)> env_block(GetEnvironmentStringsW(), free);
-
-        for (LPWCH i = env_block.get(); *i != L'\0'; ++i)
+        struct EnvironmentStringsW
         {
-            msg::write_unlocalized_text_to_stdout(Color::none, Strings::concat(Strings::to_utf8(i), '\n'));
-
-            while (*i != L'\0')
-            {
-                ++i;
-            }
+            LPWCH strings;
+            EnvironmentStringsW() : strings(GetEnvironmentStringsW()) { }
+            ~EnvironmentStringsW() { FreeEnvironmentStringsW(strings); }
+            EnvironmentStringsW(const EnvironmentStringsW&) = delete;
+        } env_block;
+
+        size_t len;
+        for (LPWCH i = env_block.strings; 0 != (len = wcslen(i)); i += len + 1)
+        {
+            result.append(Strings::to_utf8(i, len)).push_back('\n');
         }
 #else
         char** s = environ;
         for (; *s; s++)
         {
-            msg::write_unlocalized_text_to_stdout(Color::none, Strings::concat(*s, "\n"));
+            result += *s;
+            result += '\n';
```

Comment:
```suggestion
            result.append(*s).push_back('\n');
```

For consistency with the Windows version.